File: DoubleList.java
package structures;

import exceptions.IndexOutOfRangeException;

public class DoubleList extends Structure {

    private class Node {
        String data;
        Node prev;
        Node next;

        Node(String data) {
            this.data = data;
            this.prev = null;
            this.next = null;
        }
    }

    private Node head;
    private Node tail;
    private int size;

    public DoubleList() {
        super();
        head = null;
        tail = null;
        size = 0;
    }

    public DoubleList(String name) {
        super(name);
        head = null;
        tail = null;
        size = 0;
    }

    public void addHead(String value) {
        Node newNode = new Node(value);

        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.next = head;
            head.prev = newNode;
            head = newNode;
        }
        size++;
    }

    public void addTail(String value) {
        Node newNode = new Node(value);

        if (tail == null) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.prev = tail;
            tail.next = newNode;
            tail = newNode;
        }
        size++;
    }

    public void addAfter(int index, String value) throws IndexOutOfRangeException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfRangeException("Index out of range: " + index);
        }

        if (index == size - 1) {
            addTail(value);
            return;
        }

        Node current = getNodeAt(index);
        Node newNode = new Node(value);

        newNode.next = current.next;
        newNode.prev = current;

        if (current.next != null) {
            current.next.prev = newNode;
        }
        current.next = newNode;

        size++;
    }

    public void addBefore(int index, String value) throws IndexOutOfRangeException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfRangeException("Index out of range: " + index);
        }

        if (index == 0) {
            addHead(value);
            return;
        }

        Node current = getNodeAt(index);
        Node newNode = new Node(value);

        newNode.prev = current.prev;
        newNode.next = current;

        if (current.prev != null) {
            current.prev.next = newNode;
        } else {
            head = newNode;
        }
        current.prev = newNode;

        size++;
    }

    public String deleteHead() throws IndexOutOfRangeException {
        if (head == null) {
            throw new IndexOutOfRangeException("List is empty");
        }

        String value = head.data;

        if (head == tail) {
            head = null;
            tail = null;
        } else {
            head = head.next;
            head.prev = null;
        }

        size--;
        return value;
    }

    public String deleteTail() throws IndexOutOfRangeException {
        if (tail == null) {
            throw new IndexOutOfRangeException("List is empty");
        }

        String value = tail.data;

        if (head == tail) {
            head = null;
            tail = null;
        } else {
            tail = tail.prev;
            tail.next = null;
        }

        size--;
        return value;
    }

    public void deleteAt(int index) throws IndexOutOfRangeException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfRangeException("Index out of range: " + index);
        }

        if (index == 0) {
            deleteHead();
        } else if (index == size - 1) {
            deleteTail();
        } else {
            Node toDelete = getNodeAt(index);
            toDelete.prev.next = toDelete.next;
            toDelete.next.prev = toDelete.prev;
            size--;
        }
    }

    public boolean deleteByValue(String value) {
        Node current = head;
        int index = 0;

        while (current != null) {
            if (current.data.equals(value)) {
                try {
                    deleteAt(index);
                    return true;
                } catch (IndexOutOfRangeException e) {
                    return false;
                }
            }
            current = current.next;
            index++;
        }
        return false;
    }

    public String get(int index) throws IndexOutOfRangeException {
        return getNodeAt(index).data;
    }

    private Node getNodeAt(int index) throws IndexOutOfRangeException {
        if (index < 0 || index >= size) {
            throw new IndexOutOfRangeException("Index out of range: " + index);
        }

        // Optimize: traverse from head or tail based on index
        if (index <= size / 2) {
            Node current = head;
            for (int i = 0; i < index; i++) {
                current = current.next;
            }
            return current;
        } else {
            Node current = tail;
            for (int i = size - 1; i > index; i--) {
                current = current.prev;
            }
            return current;
        }
    }

    public Node findByValue(String value) {
        Node current = head;
        while (current != null) {
            if (current.data.equals(value)) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    public int getSize() {
        return size;
    }

    public void deleteNodesBeforeIndex(int index) throws IndexOutOfRangeException {
        if (index <= 0) {
            return;
        }
        if (index >= size) {
            // Удалить все узлы
            head = null;
            tail = null;
            size = 0;
            return;
        }

        // Получаем новый начальный узел
        Node newHead = getNodeAt(index);
        if (newHead != null) {
            // Разрываем связь с предыдущими узлами
            newHead.prev = null;
            
            // Освобождаем память предыдущих узлов (опционально, GC сделает это)
            Node current = head;
            while (current != newHead) {
                Node next = current.next;
                current.prev = null;
                current.next = null;
                current = next;
            }
            
            // Устанавливаем новую голову
            head = newHead;
            size -= index;  // Корректируем размер
        }
    }

    public void deleteNodesAfterIndex(int index) throws IndexOutOfRangeException {
        if (index < 0 || index >= size - 1) {
            return;
        }

        // Найти узел по индексу
        Node current = getNodeAt(index);
        if (current == null) {
            return;
        }

        // Удалить узлы после 'current'
        Node toDelete = current.next;
        while (toDelete != null) {
            Node next = toDelete.next;  // Сохраняем следующий узел
            // Очищаем связи узла
            toDelete.prev = null;
            toDelete.next = null;
            size--;
            toDelete = next;  // Переходим к следующему узлу
        }

        current.next = null;
        tail = current;
    }

    public void deleteNodesFromTo(int start, int end) throws IndexOutOfRangeException {
        if (start < 0 || end >= size || start > end) {
            throw new IndexOutOfRangeException("Invalid index range");
        }

        int deleteCount = end - start + 1;
        
        // Если удаляем с начала
        if (start == 0) {
            if (end == size - 1) {
                // Удаляем все узлы
                head = null;
                tail = null;
                size = 0;
            } else {
                // Устанавливаем новую голову
                Node newHead = getNodeAt(end + 1);
                newHead.prev = null;
                
                // Освобождаем удаляемые узлы
                Node current = head;
                while (current != newHead) {
                    Node next = current.next;
                    current.prev = null;
                    current.next = null;
                    current = next;
                }
                
                head = newHead;
                size -= deleteCount;
            }
            return;
        }

        // Если удаляем до конца
        if (end == size - 1) {
            Node newTail = getNodeAt(start - 1);
            newTail.next = null;
            
            // Освобождаем удаляемые узлы
            Node current = newTail.next;
            while (current != null) {
                Node next = current.next;
                current.prev = null;
                current.next = null;
                current = next;
            }
            
            tail = newTail;
            size = start;  // Новый размер = start (индексация с 0)
            return;
        }

        // Общий случай: удаление в середине
        Node beforeStart = getNodeAt(start - 1);
        Node afterEnd = getNodeAt(end + 1);
        
        // Устанавливаем связи
        beforeStart.next = afterEnd;
        afterEnd.prev = beforeStart;
        
        // Освобождаем удаляемые узлы
        Node current = getNodeAt(start);
        for (int i = 0; i < deleteCount && current != afterEnd; i++) {
            Node next = current.next;
            current.prev = null;
            current.next = null;
            current = next;
        }
        
        // Обновляем размер
        size -= deleteCount;
    }

    @Override
    public String serialize() {
        StringBuilder sb = new StringBuilder();
        sb.append("L ").append(name).append(" ").append(size);

        Node current = head;
        while (current != null) {
            sb.append(" ").append(current.data);
            current = current.next;
        }

        return sb.toString();
    }

    @Override
    public void deserialize(String data) {
        String[] parts = data.split(" ");
        if (parts.length < 3) {
            return;
        }

        this.name = parts[1];
        int count = Integer.parseInt(parts[2]);

        // Clear existing list
        head = null;
        tail = null;
        size = 0;

        for (int i = 0; i < count && (i + 3) < parts.length; i++) {
            addTail(parts[i + 3]);
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("DoubleList \"").append(name).append("\" (size: ").append(size).append("): [");

        Node current = head;
        while (current != null) {
            if (current != head) {
                sb.append(", ");
            }
            sb.append(current.data);
            current = current.next;
        }

        sb.append("]");
        return sb.toString();
    }
    
    // Дополнительные методы для удобства работы
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public void clear() {
        // Помогаем сборщику мусора
        Node current = head;
        while (current != null) {
            Node next = current.next;
            current.prev = null;
            current.next = null;
            current = next;
        }
        head = null;
        tail = null;
        size = 0;
    }
    
    public String getFirst() throws IndexOutOfRangeException {
        if (head == null) {
            throw new IndexOutOfRangeException("List is empty");
        }
        return head.data;
    }
    
    public String getLast() throws IndexOutOfRangeException {
        if (tail == null) {
            throw new IndexOutOfRangeException("List is empty");
        }
        return tail.data;
    }
}

File: double_list.rs
// src/double_list.rs
use crate::{Structure, AsAny};
use std::fmt;

/// Узел двусвязного списка
#[derive(Debug)]
pub struct DFNode {
    pub key: String,
    pub next: Option<Box<DFNode>>,
    pub prev: Option<*mut DFNode>, // Сырой указатель для обратной ссылки
}

impl DFNode {
    pub fn new(key: String) -> Self {
        DFNode {
            key,
            next: None,
            prev: None,
        }
    }
}

/// Двусвязный список
#[derive(Debug)]
pub struct DFList {
    pub head: Option<Box<DFNode>>,
    pub tail: Option<*mut DFNode>, // Сырой указатель на хвост
    pub length: usize,
    pub name: String,
}

impl DFList {
    /// Создает новый пустой двусвязный список
    pub fn new() -> Self {
        DFList {
            head: None,
            tail: None,
            length: 0,
            name: String::new(),
        }
    }
    
    /// Добавляет элемент в начало списка
    pub fn push_front(&mut self, key: String) {
        let mut new_node = Box::new(DFNode::new(key));
        
        match self.head.take() {
            Some(mut old_head) => {
                // Старая голова становится второй
                old_head.prev = Some(new_node.as_mut() as *mut DFNode);
                new_node.next = Some(old_head);
                self.head = Some(new_node);
            }
            None => {
                // Список был пуст
                self.tail = Some(new_node.as_mut() as *mut DFNode);
                self.head = Some(new_node);
            }
        }
        
        self.length += 1;
    }
    
    /// Добавляет элемент в конец списка
    pub fn push_back(&mut self, key: String) {
        let mut new_node = Box::new(DFNode::new(key));
        
        unsafe {
            let new_node_ptr = new_node.as_mut() as *mut DFNode;
            
            match self.tail {
                Some(mut tail_ptr) => {
                    // Присоединяем новый узел к хвосту
                    if let Some(tail) = tail_ptr.as_mut() {
                        tail.next = Some(new_node);
                        new_node.prev = Some(tail_ptr);
                        self.tail = Some(new_node_ptr);
                    }
                }
                None => {
                    // Список был пуст
                    self.head = Some(new_node);
                    self.tail = Some(new_node_ptr);
                }
            }
        }
        
        self.length += 1;
    }
    
    /// Удаляет первый элемент и возвращает его значение
    pub fn pop_front(&mut self) -> Option<String> {
        self.head.take().map(|mut old_head| {
            match old_head.next.take() {
                Some(mut new_head) => {
                    new_head.prev = None;
                    self.head = Some(new_head);
                }
                None => {
                    // Список стал пустым
                    self.tail = None;
                }
            }
            
            self.length -= 1;
            old_head.key
        })
    }
    
    /// Удаляет последний элемент и возвращает его значение
    pub fn pop_back(&mut self) -> Option<String> {
        unsafe {
            self.tail.take().map(|tail_ptr| {
                if let Some(tail) = tail_ptr.as_mut() {
                    // Находим предыдущий узел
                    match tail.prev {
                        Some(mut prev_ptr) => {
                            if let Some(prev) = prev_ptr.as_mut() {
                                prev.next = None;
                                self.tail = Some(prev_ptr);
                            }
                        }
                        None => {
                            // Это был последний/единственный элемент
                            self.head = None;
                            self.tail = None;
                        }
                    }
                }
                
                self.length -= 1;
                
                // Извлекаем значение из Box (опасно, но нужно)
                let tail_box = Box::from_raw(tail_ptr);
                tail_box.key
            })
        }
    }
    
    /// Получает элемент по индексу
    pub fn get(&self, index: usize) -> Option<&str> {
        if index >= self.length {
            return None;
        }
        
        let mut current = self.head.as_ref();
        for _ in 0..index {
            current = current.and_then(|node| node.next.as_ref());
        }
        
        current.map(|node| node.key.as_str())
    }
    
    /// Устанавливает элемент по индексу
    pub fn set(&mut self, index: usize, value: String) -> Result<(), String> {
        if index >= self.length {
            return Err("Index out of bounds".to_string());
        }
        
        let mut current = self.head.as_mut();
        for _ in 0..index {
            current = current.and_then(|node| node.next.as_mut());
        }
        
        if let Some(node) = current {
            node.key = value;
            Ok(())
        } else {
            Err("Node not found".to_string())
        }
    }
    
    /// Вставляет элемент перед заданным индексом
    pub fn insert_before(&mut self, index: usize, key: String) -> Result<(), String> {
        if index > self.length {
            return Err("Index out of bounds".to_string());
        }
        
        if index == 0 {
            self.push_front(key);
            return Ok(());
        }
        
        if index == self.length {
            self.push_back(key);
            return Ok(());
        }
        
        // Находим узел, перед которым вставляем
        let mut current = self.head.as_mut();
        for _ in 0..index {
            current = current.and_then(|node| node.next.as_mut());
        }
        
        if let Some(current_node) = current {
            let mut new_node = Box::new(DFNode::new(key));
            
            unsafe {
                let new_node_ptr = new_node.as_mut() as *mut DFNode;
                
                // Связываем с предыдущим узлом
                if let Some(prev_ptr) = current_node.prev {
                    if let Some(prev) = prev_ptr.as_mut() {
                        prev.next = Some(new_node);
                        new_node.prev = Some(prev_ptr);
                    }
                }
                
                // Связываем с текущим узлом
                new_node.next = Some(Box::from_raw(current_node as *mut DFNode));
                current_node.prev = Some(new_node_ptr);
            }
            
            self.length += 1;
            Ok(())
        } else {
            Err("Node not found".to_string())
        }
    }
    
    /// Вставляет элемент после заданного индекса
    pub fn insert_after(&mut self, index: usize, key: String) -> Result<(), String> {
        if index >= self.length {
            return Err("Index out of bounds".to_string());
        }
        
        if index == self.length - 1 {
            self.push_back(key);
            return Ok(());
        }
        
        // Находим узел, после которого вставляем
        let mut current = self.head.as_mut();
        for _ in 0..index {
            current = current.and_then(|node| node.next.as_mut());
        }
        
        if let Some(current_node) = current {
            let mut new_node = Box::new(DFNode::new(key));
            
            unsafe {
                let new_node_ptr = new_node.as_mut() as *mut DFNode;
                
                // Связываем со следующим узлом
                if let Some(mut next_node) = current_node.next.take() {
                    next_node.prev = Some(new_node_ptr);
                    new_node.next = Some(next_node);
                }
                
                // Связываем с текущим узлом
                current_node.next = Some(new_node);
                new_node.prev = Some(current_node as *mut DFNode);
            }
            
            self.length += 1;
            Ok(())
        } else {
            Err("Node not found".to_string())
        }
    }
    
    /// Удаляет элемент по индексу
    pub fn remove(&mut self, index: usize) -> Result<String, String> {
        if index >= self.length {
            return Err("Index out of bounds".to_string());
        }
        
        if index == 0 {
            return self.pop_front().ok_or_else(|| "List is empty".to_string());
        }
        
        if index == self.length - 1 {
            return self.pop_back().ok_or_else(|| "List is empty".to_string());
        }
        
        // Находим удаляемый узел
        let mut current = self.head.as_mut();
        for _ in 0..index {
            current = current.and_then(|node| node.next.as_mut());
        }
        
        if let Some(node_to_remove) = current {
            let value = node_to_remove.key.clone();
            
            unsafe {
                // Связываем предыдущий и следующий узлы
                if let Some(prev_ptr) = node_to_remove.prev {
                    if let Some(prev) = prev_ptr.as_mut() {
                        prev.next = node_to_remove.next.take();
                    }
                }
                
                if let Some(mut next_node) = node_to_remove.next.take() {
                    next_node.prev = node_to_remove.prev;
                }
            }
            
            self.length -= 1;
            Ok(value)
        } else {
            Err("Node not found".to_string())
        }
    }
    
    /// Удаляет первый узел с заданным значением
    pub fn remove_by_value(&mut self, value: &str) -> bool {
        let mut current = self.head.as_mut();
        let mut index = 0;
        
        while let Some(node) = current {
            if node.key == value {
                if self.remove(index).is_ok() {
                    return true;
                }
            }
            current = node.next.as_mut();
            index += 1;
        }
        
        false
    }
    
    /// Находит узел с заданным значением
    pub fn find(&self, value: &str) -> Option<usize> {
        let mut current = self.head.as_ref();
        let mut index = 0;
        
        while let Some(node) = current {
            if node.key == value {
                return Some(index);
            }
            current = node.next.as_ref();
            index += 1;
        }
        
        None
    }
    
    /// Проверяет, пуст ли список
    pub fn is_empty(&self) -> bool {
        self.length == 0
    }
    
    /// Возвращает длину списка
    pub fn len(&self) -> usize {
        self.length
    }
    
    /// Очищает список
    pub fn clear(&mut self) {
        self.head = None;
        self.tail = None;
        self.length = 0;
    }
    
    /// Преобразует список в вектор
    pub fn to_vec(&self) -> Vec<String> {
        let mut result = Vec::with_capacity(self.length);
        let mut current = self.head.as_ref();
        
        while let Some(node) = current {
            result.push(node.key.clone());
            current = node.next.as_ref();
        }
        
        result
    }
}

impl Structure for DFList {
    fn serialize(&self) -> String {
        let elements: Vec<String> = self.to_vec();
        format!("L {} {} {}", self.name, self.length, elements.join(" "))
    }
    
    fn deserialize(&mut self, data: &str) {
        let parts: Vec<&str> = data.split_whitespace().collect();
        
        if parts.len() < 3 || parts[0] != "L" {
            return;
        }
        
        self.name = parts[1].to_string();
        self.length = parts[2].parse().unwrap_or(0);
        
        // Очищаем текущий список
        self.clear();
        
        // Добавляем элементы
        for i in 0..self.length {
            if let Some(value) = parts.get(3 + i) {
                self.push_back(value.to_string());
            }
        }
    }
    
    fn name(&self) -> &str {
        &self.name
    }
    
    fn set_name(&mut self, name: String) {
        self.name = name;
    }
}

//impl AsAny for DFList {
//    fn as_any(&self) -> &dyn std::any::Any {
//        self
//    }
//    
//    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
//        self
//    }
//}

impl Default for DFList {
    fn default() -> Self {
        Self::new()
    }
}

impl fmt::Display for DFList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "DoubleList[")?;
        
        let mut current = self.head.as_ref();
        let mut first = true;
        
        while let Some(node) = current {
            if !first {
                write!(f, ", ")?;
            }
            write!(f, "{}", node.key)?;
            first = false;
            current = node.next.as_ref();
        }
        
        write!(f, "]")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_push_front() {
        let mut list = DFList::new();
        list.push_front("world".to_string());
        list.push_front("hello".to_string());
        
        assert_eq!(list.len(), 2);
        assert_eq!(list.get(0), Some("hello"));
        assert_eq!(list.get(1), Some("world"));
    }
    
    #[test]
    fn test_push_back() {
        let mut list = DFList::new();
        list.push_back("hello".to_string());
        list.push_back("world".to_string());
        
        assert_eq!(list.len(), 2);
        assert_eq!(list.get(0), Some("hello"));
        assert_eq!(list.get(1), Some("world"));
    }
    
    #[test]
    fn test_pop_front() {
        let mut list = DFList::new();
        list.push_back("hello".to_string());
        list.push_back("world".to_string());
        
        assert_eq!(list.pop_front(), Some("hello".to_string()));
        assert_eq!(list.len(), 1);
        assert_eq!(list.get(0), Some("world"));
    }
    
    #[test]
    fn test_pop_back() {
        let mut list = DFList::new();
        list.push_back("hello".to_string());
        list.push_back("world".to_string());
        
        assert_eq!(list.pop_back(), Some("world".to_string()));
        assert_eq!(list.len(), 1);
        assert_eq!(list.get(0), Some("hello"));
    }
    
    #[test]
    fn test_insert_and_remove() {
        let mut list = DFList::new();
        list.push_back("a".to_string());
        list.push_back("c".to_string());
        
        list.insert_before(1, "b".to_string()).unwrap();
        assert_eq!(list.len(), 3);
        assert_eq!(list.get(1), Some("b"));
        
        let removed = list.remove(1).unwrap();
        assert_eq!(removed, "b");
        assert_eq!(list.len(), 2);
    }
    
    #[test]
    fn test_serialize() {
        let mut list = DFList::new();
        list.set_name("test_list".to_string());
        list.push_back("hello".to_string());
        list.push_back("world".to_string());
        
        let serialized = list.serialize();
        assert_eq!(serialized, "L test_list 2 hello world");
    }
    
    #[test]
    fn test_deserialize() {
        let mut list = DFList::new();
        list.deserialize("L my_list 3 a b c");
        
        assert_eq!(list.name(), "my_list");
        assert_eq!(list.len(), 3);
        assert_eq!(list.get(0), Some("a"));
        assert_eq!(list.get(1), Some("b"));
        assert_eq!(list.get(2), Some("c"));
    }
}
